hathor/healthcheck/resources/healthcheck.py

```diff
@@ -0,0 +1,172 @@
+import hathor
+from hathor.api_util import Resource
+from hathor.cli.openapi_files.register import register_resource
+from hathor.healthcheck.models import ComponentHealthCheck, ComponentType, HealthCheckStatus, ServiceHealthCheck
+from hathor.manager import HathorManager
+from hathor.util import json_dumpb
+
+
+def build_sync_health_status(manager: HathorManager) -> ComponentHealthCheck:
+    """Builds the sync health status object."""
+    healthy, reason = manager.is_sync_healthy()
+
+    return ComponentHealthCheck(
+        component_name="sync",
+        component_type=ComponentType.INTERNAL,
+        status=HealthCheckStatus.PASS if healthy else HealthCheckStatus.FAIL,
+        output=reason or "Healthy",
+    )
+
+
+@register_resource
+class HealthcheckResource(Resource):
+    isLeaf = True
+
+    def __init__(self, manager: HathorManager):
+        self.manager = manager
+
+    def render_GET(self, request):
+        """ GET request /health/
+            Returns the health status of the fullnode
+
+            :rtype: string (json)
+        """
+        components_health_checks = [
+            build_sync_health_status(self.manager)
+        ]
+
+        health_check = ServiceHealthCheck(
+            description=f"Hathor-core {hathor.__version__}",
+            checks={c.component_name: [c] for c in components_health_checks},
+        )
+
+        request.setResponseCode(health_check.get_http_status_code())
+        return json_dumpb(health_check.to_json())
+
+
+# TODO: Fix below
+HealthcheckResource.openapi = {
+    '/health': {
+        'x-visibility': 'public',
+        'x-rate-limit': {
+            'global': [
+                {
+                    'rate': '10r/s',
+                    'burst': 10,
+                    'delay': 5
+                }
+            ],
+            'per-ip': [
+                {
+                    'rate': '1r/s',
+                    'burst': 3,
+                    'delay': 2
+                }
+            ]
+        },
+        'get': {
+            'tags': ['healthcheck'],
+            'operationId': 'get',
+            'summary': 'Health status of the fullnode',
+            'description': '''
+Returns 200 if the fullnode should be considered healthy.
+
+Returns 503 otherwise. The response will contain the components that were considered for the healthcheck
+and the reason why they were unhealthy.
```

Comment:
Is there any particular motivation for using 503? My intuition would be that the direct response from the fullnode would always be 200 if it was able to compute the request and the content of the request would have the health status. I'd expect 503 to be returned by a proxy that couldn't reach the fullnode, in case we use nginx (which we typically do) and the request timeouts, it will respond with a 503, however if the fullnode also responds with 503 normally we would only be able to differentiate by looking at the content (json response from fullnode vs html response from nginx).

If this is by design then I think the motivation should be commented more clearly. Otherwise it seems better to just use 200 and let the caller look at the content. As in, 200 always means the response came from the fullnode and has a predictable json format.

Reply:
Most infrastructural tooling that work with health checks expect them to work based on HTTP codes, so returning any code > 399 facilitates integration with them.

Even the uptime monitoring tools that we evaluated to use do a better job with HTTP codes than reading the response code. Most of them have at most a basic support to parsing the response code that usually just searches for the presence of a string on them, if at all.

AWS Load Balancers themselves are another instance of this. Their health check works solely based on HTTP codes.

So I think we should stick to this to have an easier integration with any such tool. But I could change the code from 503 to any other that is > 399.

What do you think?

Reply:
It looks like 503 would be useful in practice then. If in practice, the content of the response isn't important when the status is 503, I'm fine using it as is and just commenting that we use 503 because it integrates better with external tools.

Otherwise, if the content ends up being used in practice it could be useful to have some way of signaling (possibly with a GET parameter like `?strict_status_code=1`) that we want a code 200 with even with an error response (or the other way around depending on what we choose as a default), that way the tools that do use the content can more easily expect 200 for knowing they can parse the response content.

Reply:
I think it's interesting to allow choosing the behavior. I'll implement that.

Reply:
Done in https://github.com/HathorNetwork/hathor-core/pull/807/commits/456a63bd3125b248fece5ea53259774633b60349